File: Practical Asynchronous Programming for VS2012 or higher/[C#][VB]-Practical Asynchronous Programming for VS2012 or higher/C# and VB/WindowsFormsApplication1/Form1.cs
using DataOperations;
using System;
using System.Data;
using System.Threading;
using System.Threading.Tasks;
using System.Windows.Forms;

namespace WindowsFormsApplication1
{
    /// <summary>
    /// Contains code to return simple data from a ms-access table
    /// using Task-based Asynchronous Pattern in a class project
    /// </summary>
    public partial class Form1 : Form
    {
        /// <summary>
        /// At form level for cancel capability
        /// </summary>
        private CancellationTokenSource cts;
       
        public Form1()
        {
            InitializeComponent();
            this.Shown += Form1_Shown;
        }

        void Form1_Shown(object sender, EventArgs e)
        {
            ActiveControl = cmdGetData;
        }

        private void cmdGetData_Click(object sender, EventArgs e)
        {
            ResetData();
            Work w = new Work();
            w.Caller = this;
            w.Label = label1;

            dataGridView1.DataSource = w.dtPeople;

            cts = new CancellationTokenSource();
            Task.Run(() => w.GetDataAsync(cts.Token));

        }
        private void cmdCancel_Click(object sender, EventArgs e)
        {
            cts.Cancel();
        }
        private void ResetData()
        {
            label1.Text = "0";
            label1.Refresh();

            if (cts != null)
            {
                cts.Cancel();
            }

            if (dataGridView1.DataSource != null)
            {
                var dt = (DataTable)dataGridView1.DataSource;
                dt.Rows.Clear();
                dataGridView1.Refresh();

                Thread.Sleep(1000);
            }
        }

    }
}

File: Practical Asynchronous Programming for VS2012 or higher/[C#][VB]-Practical Asynchronous Programming for VS2012 or higher/C# and VB/WindowsApplication2/Form1.vb
Imports System.Threading

Public Class Form1
    Private ct As CancellationTokenSource

    Private Sub Form1_Load(sender As Object, e As EventArgs) Handles MyBase.Load
        ActiveControl = cmdGetData
    End Sub

    Private Sub cmdGetData_Click(sender As Object, e As EventArgs) Handles cmdGetData.Click
        ResetData()
        Dim w As New DataOperations()
        w.Caller = Me
        w.Label = label1

        dataGridView1.DataSource = w.DataTable

        ct = New CancellationTokenSource()
        Task.Run(Function() w.GetDataAsync(ct.Token))
    End Sub
    Private Sub ResetData()
        label1.Text = "0"
        label1.Refresh()

        If ct IsNot Nothing Then
            ct.Cancel()
        End If

        If dataGridView1.DataSource IsNot Nothing Then
            Dim dt = CType(dataGridView1.DataSource, DataTable)
            dt.Rows.Clear()
            dataGridView1.Refresh()

            Thread.Sleep(1000)
        End If
    End Sub

    Private Sub cmdCancel_Click(sender As Object, e As EventArgs) Handles cmdCancel.Click
        ct.Cancel()
    End Sub
End Class
